pkg/kubelet/kubelet.go

```diff
@@ -171,83 +172,154 @@ type SyncHandler interface {
 // Option is a functional option type for Kubelet
 type Option func(*Kubelet)
 
+// bootstrapping interface for kubelet, targets the initialization protocol
+type KubeletBootstrap interface {
+	GetConfig() *KubeletConfig
+	BirthCry()
+	StartGarbageCollection()
+	ListenAndServe(address net.IP, port uint, tlsOptions *server.TLSOptions, auth server.AuthInterface, enableDebuggingHandlers bool)
+	ListenAndServeReadOnly(address net.IP, port uint)
+	Run(<-chan kubetypes.PodUpdate)
+	RunOnce(<-chan kubetypes.PodUpdate) ([]RunPodResult, error)
+}
+
+// create and initialize a Kubelet instance
+type KubeletBuilder func(kc_old *KubeletConfig, kc_new *componentconfig.KubeletConfiguration) (KubeletBootstrap, error)
+
+// KubeletConfig is all of the parameters necessary for running a kubelet.
+// TODO: This should probably be merged with KubeletServer.  The extra object is a consequence of refactoring.
+type KubeletConfig struct {
+	Auth                    server.AuthInterface
+	AutoDetectCloudProvider bool
+	Builder                 KubeletBuilder
+	CAdvisorInterface       cadvisor.Interface
+	Cloud                   cloudprovider.Interface
+	ContainerManager        cm.ContainerManager
+	DockerClient            dockertools.DockerInterface
+	EventClient             *clientset.Clientset
+	Hostname                string
+	KubeClient              *clientset.Clientset
+	Mounter                 mount.Interface
+	NetworkPlugins          []network.NetworkPlugin
+	NodeName                string
+	OOMAdjuster             *oom.OOMAdjuster
+	OSInterface             kubecontainer.OSInterface
+	PodConfig               *config.PodConfig
+	Recorder                record.EventRecorder
+	TLSOptions              *server.TLSOptions
+	Writer                  kubeio.Writer
+	VolumePlugins           []volume.VolumePlugin
+	ContainerRuntimeOptions []kubecontainer.Option
+	Options                 []Option
+}
+
+func makePodSourceConfig(kc_old *KubeletConfig, kc_new *componentconfig.KubeletConfiguration) (*config.PodConfig, error) {
```

Comment:
why does this func accept pointers to config args? will they be mutated or are they in a performance sensitive call chain? doesn't look that way to me